File: generator/src/test/java/quan/generator/role/SRoleLogin.java
package quan.generator.role;

import java.util.HashSet;
import quan.generator.user.UserInfo;
import quan.network.message.Buffer;
import java.io.IOException;
import java.util.HashMap;
import quan.network.message.Message;
import java.util.ArrayList;

/**
 * 角色登录
 * Created by 自动生成
 */
public class SRoleLogin extends Message {

    private long roleId;//角色id
    private RoleInfo roleInfo;//角色信息
    private ArrayList<RoleInfo> roleInfoList;//角色信息
    private HashSet<RoleInfo> roleInfoSet;//角色信息
    private HashMap<Long, RoleInfo> roleInfoMap;//角色信息
    private UserInfo userInfo;//用户信息

    public SRoleLogin() {
        super(2222);
        roleId = 111L;
        roleInfo = new RoleInfo();
        roleInfoList = new ArrayList<>();
        roleInfoSet = new HashSet<>();
        roleInfoMap = new HashMap<>();
        userInfo = new UserInfo();
    }

    public long getRoleId() {
        return roleId;
    }

    public void setRoleId(long roleId) {
        this.roleId = roleId;
    }

    public RoleInfo getRoleInfo() {
        return roleInfo;
    }

    public void setRoleInfo(RoleInfo roleInfo) {
        this.roleInfo = roleInfo;
    }

    public ArrayList<RoleInfo> getRoleInfoList() {
        return roleInfoList;
    }

    public HashSet<RoleInfo> getRoleInfoSet() {
        return roleInfoSet;
    }

    public HashMap<Long, RoleInfo> getRoleInfoMap() {
        return roleInfoMap;
    }

    public UserInfo getUserInfo() {
        return userInfo;
    }

    public void setUserInfo(UserInfo userInfo) {
        this.userInfo = userInfo;
    }


    @Override
    public SRoleLogin create() {
        return new SRoleLogin();
    }

    @Override
    public void encode(Buffer buffer) throws IOException {
        super.encode(buffer);
        buffer.writeLong(roleId);
        roleInfo.encode(buffer);
        buffer.writeInt(roleInfoList.size());
        for (RoleInfo roleInfoListValue : roleInfoList) {
            roleInfoListValue.encode(buffer);
        }
        buffer.writeInt(roleInfoSet.size());
        for (RoleInfo roleInfoSetValue : roleInfoSet) {
            roleInfoSetValue.encode(buffer);
        }
        buffer.writeInt(roleInfoMap.size());
        for (long roleInfoMapKey : roleInfoMap.keySet()) {
            buffer.writeLong(roleInfoMapKey);
            roleInfoMap.get(roleInfoMapKey).encode(buffer);
        }
        userInfo.encode(buffer);
    }

    @Override
    public void decode(Buffer buffer) throws IOException {
        super.encode(buffer);
        roleId = buffer.readLong();
        roleInfo.decode(buffer);
        int roleInfoListSize = buffer.readInt();
        for (int i = 0; i < roleInfoListSize; i++) {
            RoleInfo roleInfoListValue = new RoleInfo();
            roleInfoListValue.decode(buffer);
            roleInfoList.add(roleInfoListValue);
        }
        int roleInfoSetSize = buffer.readInt();
        for (int i = 0; i < roleInfoSetSize; i++) {
            RoleInfo roleInfoSetValue = new RoleInfo();
            roleInfoSetValue.decode(buffer);
            roleInfoSet.add(roleInfoSetValue);
        }
        int roleInfoMapSize = buffer.readInt();
        for (int i = 0; i < roleInfoMapSize; i++) {
            long roleInfoMapKey = buffer.readLong();
            RoleInfo roleInfoMapValue = new RoleInfo();
            roleInfoMapValue.decode(buffer);
            roleInfoMap.put(roleInfoMapKey, roleInfoMapValue);
        }
        userInfo.decode(buffer);
    }

    @Override
    public String toString() {
        return "SRoleLogin{" +
                "roleId=" + roleId +
                ",roleInfo=" + roleInfo +
                ",roleInfoList=" + roleInfoList +
                ",roleInfoSet=" + roleInfoSet +
                ",roleInfoMap=" + roleInfoMap +
                ",userInfo=" + userInfo +
                '}';

    }

}

File: message/message_cs/test/role/SRoleLogin.cs
using System;
using System.Collections.Generic;
using message_cs;
using Buffer = message_cs.Buffer;
using message_cs.test.user;

namespace message_cs.test.role
{
	/// <summary>
	/// 角色登录<br/>
	/// Created by 自动生成
	/// </summary>
    public class SRoleLogin : Message
    {
		public long roleId { get; set; }

		private string _roleName = "";

		public string roleName
		{
	    	get => _roleName;
	    	set => _roleName = value ?? throw new NullReferenceException();
		}

		private RoleInfo _roleInfo = new RoleInfo();

		public RoleInfo roleInfo
		{
	    	get => _roleInfo;
	    	set => _roleInfo = value ?? throw new NullReferenceException();
		}

		public List<RoleInfo> roleInfoList { get; } = new List<RoleInfo>();

		public HashSet<RoleInfo> roleInfoSet { get; } = new HashSet<RoleInfo>();

		public Dictionary<long, RoleInfo> roleInfoMap { get; } = new Dictionary<long, RoleInfo>();

		private UserInfo userInfo { get; set; }


		public SRoleLogin(): base(222)
		{
		}

		public override Message Create()
		{
			return new SRoleLogin();
		}

		public override void Encode(Buffer buffer)
		{
	    	base.Encode(buffer);

		    buffer.WriteLong(roleId);
		    buffer.WriteString(roleName);
		    roleInfo.Encode(buffer);

		    buffer.WriteInt(roleInfoList.Count);
		    foreach (var _roleInfoList_Value in roleInfoList) {
			    _roleInfoList_Value.Encode(buffer);
		    }

		    buffer.WriteInt(roleInfoSet.Count);
		    foreach (var _roleInfoSet_Value in roleInfoSet) {
			    _roleInfoSet_Value.Encode(buffer);
		    }

		    buffer.WriteInt(roleInfoMap.Count);
		    foreach (var _roleInfoMap_Key in roleInfoMap.Keys) {
		        buffer.WriteLong(_roleInfoMap_Key);
			    roleInfoMap[_roleInfoMap_Key].Encode(buffer);
		    }

		    buffer.WriteBool(userInfo != null);
		    userInfo?.Encode(buffer);

		}

		public override void Decode(Buffer buffer)
		{
	    	base.Decode(buffer);

		    roleId = buffer.ReadLong();
		    roleName = buffer.ReadString();
		    roleInfo.Decode(buffer);

		    var _roleInfoList_Size = buffer.ReadInt();
		    for (var _index_ = 0; _index_ < _roleInfoList_Size; _index_++) {
			    var _roleInfoList_Value = new RoleInfo();
			    _roleInfoList_Value.Decode(buffer);
			    roleInfoList.Add(_roleInfoList_Value);
		    }

		    var _roleInfoSet_Size = buffer.ReadInt();
		    for (var _index_ = 0; _index_ < _roleInfoSet_Size; _index_++) {
			    var _roleInfoSet_Value = new RoleInfo();
			    _roleInfoSet_Value.Decode(buffer);
			    roleInfoSet.Add(_roleInfoSet_Value);
		    }

		    var _roleInfoMap_Size = buffer.ReadInt();
		    for (var _index_ = 0; _index_ < _roleInfoMap_Size; _index_++) {
			    var _roleInfoMap_Key = buffer.ReadLong();
			    var _roleInfoMap_Value = new RoleInfo();
			    _roleInfoMap_Value.Decode(buffer);
			    roleInfoMap.Add(_roleInfoMap_Key, _roleInfoMap_Value);
		    }

		    if (buffer.ReadBool()) {
		        if (userInfo == null) {
		            userInfo = new UserInfo();
		        }
		        userInfo.Decode(buffer);
            }

		}

		public override string ToString()
		{
			return "SRoleLogin{" +
					"roleId=" + roleId +
					",roleName='" + roleName + '\'' +
					",roleInfo=" + roleInfo +
					",roleInfoList=" + ToString(roleInfoList) +
					",roleInfoSet=" + ToString(roleInfoSet) +
					",roleInfoMap=" + ToString(roleInfoMap) +
					",userInfo=" + userInfo +
					'}';
		}
    }
}